### PR TITLE
Add istio proxy metric scrape support

### PR DESCRIPTION
**Pull request by hisarbalik**
## Description

Changes proposed in this pull request:

- Add new metric pipeline input configuration `istio enabled`
- Add prometheus metric scrape configuration for istio proxy

## Related Issues and Documents

Closes:

Related issues:
[Istio-proxy scraping support#17887](https://github.com/kyma-project/kyma/issues/17887)

## Traceability

- [x] The PR is linked to a GitHub Issue.
- [ ] New features have a milestone label set.
- [x] New features have defined acceptance criteria in a corresponding GitHub Issue, and all criteria are satisfied with this PR.
- [x] The corresponding GitHub Issue has a respective `area` label.
- [ ] The follow-up issues (if any) are linked in the `Related Issues` section.

## Testability

The feature is unit-tested:

- [x] Yes.
- [ ] No, because unit tests are not needed.
- [ ] No, because of ...

The feature is e2e-tested:

- [ ] Yes.
- [ ] No, because e2e-tests are not needed.
- [x] No, because of ...

Tests conducted for the PR:

## Codebase

- [x] My code follows the style guidelines of this project.
- [x] The code was planned and designed following the defined architecture and the separation of concerns.
- [x] The code has sufficient comments, particularly for all hard-to-understand areas.
- [x] This PR adds value and shows no feature creep.
- [ ] I have augmented the test suite that proves my fix is effective or that my feature works.
- [ ] Adjusted the documentation if the change is user-facing.

### CODE REVIEW

File: internal/otelcollector/config/metric/agent/receivers.go [hunk 1]
```diff
@@ -138,3 +142,66 @@ func makePrometheusAppPodsConfig() *PrometheusReceiver {
...
+func makePrometheusIstioConfig() *PrometheusReceiver {
+	return &PrometheusReceiver{
+		Config: promconfig.Config{
+			ScrapeConfigs: []*promconfig.ScrapeConfig{
+				{
+					JobName:     "istio-proxy",
+					MetricsPath: "/stats/prometheus",
+					ServiceDiscoveryConfigs: []promdiscovery.Config{
+						&promk8sdiscovery.SDConfig{
+							Role:             promk8sdiscovery.RoleEndpoint,
```
Comment (a-thaler) [inline internal/otelcollector/config/metric/agent/receivers.go:155]: I think we should do a pod based discovery and not make it based on endpoint based discovery. With that we will discover also pods not being behind a service.

File: internal/otelcollector/config/metric/agent/receivers.go [hunk 2]
```diff
@@ -138,3 +142,66 @@ func makePrometheusAppPodsConfig() *PrometheusReceiver {
...
+						{
+							SourceLabels: []prommodel.LabelName{"__meta_kubernetes_pod_name"},
+							Action:       promlabel.Replace,
+							TargetLabel:  "pod",
+						},
+						{
+							SourceLabels: []prommodel.LabelName{"__meta_kubernetes_pod_node_name"},
+							Action:       promlabel.Replace,
+							TargetLabel:  "node",
+						},
```
Comment (a-thaler) [inline internal/otelcollector/config/metric/agent/receivers.go:194]: Can you also check in what state a pod should be in order to be scrapeable, see __meta_kubernetes_pod_ready and __meta_kubernetes_pod_phase labels https://prometheus.io/docs/prometheus/latest/configuration/configuration/#pod

I haven't considered that yet, might be that we explicitly need to check that it is not in "Completed" state?

Comment (a-thaler) [reply internal/otelcollector/config/metric/agent/receivers.go:194]: Seeing this one here: https://raw.githubusercontent.com/istio/istio/release-1.8/samples/addons/prometheus.yaml
It looks like we should add something like:
```
     - action: drop
        regex: Pending|Succeeded|Failed
        source_labels:
        - __meta_kubernetes_pod_phase
```
Please check that also for the annotation based pod scraping, same situation

File: internal/otelcollector/config/metric/agent/receivers.go [hunk 3]
```diff
@@ -138,3 +142,66 @@ func makePrometheusAppPodsConfig() *PrometheusReceiver {
...
+							Action:       promlabel.Keep,
+							Regex:        promlabel.MustNewRegexp("http-envoy-prom"),
+						},
+						{
+							SourceLabels: []prommodel.LabelName{"__meta_kubernetes_namespace"},
+							Action:       promlabel.Replace,
+							TargetLabel:  "namespace",
+						},
+						{
+							SourceLabels: []prommodel.LabelName{"__meta_kubernetes_service_name"},
```
[the remark this answers is not in the document]
Comment (hisarbalik) [reply internal/otelcollector/config/metric/agent/receivers.go]: removed after pod based discovery configured

File: internal/otelcollector/config/metric/agent/receivers.go (hunk 1 above)
Comment (hisarbalik) [reply internal/otelcollector/config/metric/agent/receivers.go]: discovery mode switched to pod role

File: internal/otelcollector/config/metric/agent/receivers.go (hunk 2 above)
Comment (hisarbalik) [reply internal/otelcollector/config/metric/agent/receivers.go:194]: pod phase check rule added

**Comment by chrkl** [comment]
Please make the following changes:
- Extend the decision when the metric agent DaemonSet is required [here](https://github.com/kyma-project/telemetry-manager/blob/main/internal/reconciler/metricpipeline/reconciler.go#L264)
- Add unit tests for the config builder the validate the configuration also in the case that Istio scraping is enabled
- Fill the PR template

**Comment by hisarbalik** [comment]
> Please make the following changes:
> 
> * Extend the decision when the metric agent DaemonSet is required [here](https://github.com/kyma-project/telemetry-manager/blob/main/internal/reconciler/metricpipeline/reconciler.go#L264)
> * Add unit tests for the config builder the validate the configuration also in the case that Istio scraping is enabled
> * Fill the PR template

Code already extended accordingly new requirement and unit test are already provided

File: config/development/telemetry.kyma-project.io_metricpipelines.yaml [hunk 4]
```diff
@@ -49,6 +49,13 @@ spec:
                   application:
                     description: Configures application related scraping.
                     properties:
+                      istio:
+                        description: Configures istio scraping.
```
Comment (a-thaler) [inline config/development/telemetry.kyma-project.io_metricpipelines.yaml:53]: I think the description can be extended much more. Be short but precise on what will happen when the feature is in use.

**Review by chrkl — DISMISSED** [review]

File: apis/telemetry/v1alpha1/metricpipeline_types.go [hunk 5]
```diff
@@ -89,7 +91,7 @@ type MetricPipelineContainerRuntimeInput struct {
 
 // MetricPipelineIstioInput defines the Istio scraping section.
 type MetricPipelineIstioInput struct {
-	// Indicates if Istio scraping is enabled.
+	// If enabled, metric for istio-proxy containers will be scraped from pods that have had the istio-proxy sidecar injected.
```
Comment (NHingerl) [inline apis/telemetry/v1alpha1/metricpipeline_types.go:94]:
```suggestion
	// If enabled, metrics for istio-proxy containers are scraped from Pods that have had the istio-proxy sidecar injected.
```

**Review by chrkl — APPROVED** [review]

**Review by NHingerl — APPROVED** [review]

**Pull request merged**